crates/extensions/c8y_mapper_ext/src/entity_cache.rs

```diff
@@ -0,0 +1,353 @@
+            pending_entities: PendingEntityStore::new(mqtt_schema, telemetry_cache_size),
+            external_id_mapper_fn: Box::new(external_id_mapper_fn),
+            external_id_validator_fn: Box::new(external_id_validator_fn),
+        }
+    }
+
+    pub(crate) fn register_entity(
+        &mut self,
+        entity: EntityRegistrationMessage,
+    ) -> Result<Vec<PendingEntityData>, Error> {
+        let parent = entity.parent.as_ref().unwrap_or(&self.main_device_tid);
+        if self.entities.contains_key(parent) {
+            let outcome = self.register_single_entity(entity.clone())?;
+            if outcome == InsertOutcome::Unchanged {
+                return Ok(vec![]);
+            }
+
+            let topic_id = entity.topic_id.clone();
+            let current_entity_data = self
+                .pending_entities
+                .take_cached_entity_data(entity.clone());
+            let mut pending_entities = vec![current_entity_data];
+
+            let pending_children = self
+                .pending_entities
+                .take_cached_child_entities_data(&topic_id);
+            for pending_child in pending_children {
+                let child_reg_message = pending_child.reg_message.clone();
+                self.register_single_entity(child_reg_message)?;
+                pending_entities.push(pending_child);
+            }
+            Ok(pending_entities)
+        } else {
+            self.pending_entities
+                .cache_early_registration_message(entity);
+            Ok(vec![])
+        }
+    }
+
+    pub fn register_single_entity(
+        &mut self,
+        entity: EntityRegistrationMessage,
+    ) -> Result<InsertOutcome, InvalidExternalIdError> {
+        let external_id = if let Some(id) = entity.external_id {
+            (self.external_id_validator_fn)(id.as_ref())?
+        } else {
+            (self.external_id_mapper_fn)(&entity.topic_id, self.main_device_external_id())
+        };
+
+        let parent = match entity.r#type {
+            EntityType::MainDevice => None,
+            EntityType::ChildDevice => entity
+                .parent
+                .clone()
+                .or_else(|| Some(self.main_device_tid.clone())),
+            EntityType::Service => entity
+                .parent
+                .clone()
+                .or_else(|| entity.topic_id.default_service_parent_identifier())
+                .or_else(|| Some(self.main_device_tid.clone())),
+        };
+
+        let entity_metadata = EntityMetadata {
+            topic_id: entity.topic_id.clone(),
+            external_id: Some(external_id.clone()),
+            r#type: entity.r#type,
+            parent,
+            other: entity.other,
+            twin_data: Map::new(),
+        };
+
+        let outcome = self.insert(entity.topic_id.clone(), entity_metadata);
+        self.external_id_map.insert(external_id, entity.topic_id);
```

Comment:
Updating the `external_id_map` must be done *inside* `self.insert()`.

Indeed, the `insert` method is `pub` and can possibly be invoked from elsewhere, breaking the consistency between the `entities` map and its `external_id_map` index.